server/market/market.go

```diff
@@ -247,19 +348,46 @@ func (m *Market) TxMonitored(user account.AccountID, asset uint32, txid string)
 // When submission is completed, an error value will be sent on the channel.
 // This is the asynchronous version of SubmitOrder.
 func (m *Market) SubmitOrderAsync(rec *orderRecord) <-chan error {
+	sendErr := func(err error) <-chan error {
+		errChan := make(chan error, 1)
+		errChan <- err // i.e. ErrInvalidOrder, ErrInvalidCommitment
+		return errChan
+	}
+
 	// Validate the order. The order router must do it's own validation, but do
 	// a second validation for (1) this Market and (2) epoch status, before
 	// putting it on the queue.
 	if err := m.validateOrder(rec.order); err != nil {
 		// Order ID may not be computed since ServerTime has not been set.
 		log.Debugf("SubmitOrderAsync: Invalid order received: %x: %v", rec.order.Serialize(), err)
-		errChan := make(chan error, 1)
-		errChan <- err // i.e. ErrInvalidOrder, ErrInvalidCommitment
-		return errChan
+		return sendErr(err)
+	}
+
+	trySend := func(sig *orderUpdateSignal) {
+		select {
+		case m.orderRouter <- sig:
+		case <-time.After(5 * time.Second):
+			log.Errorf("Timeout relaying incoming order %v", sig.rec.order.ID())
+			sig.errChan <- ErrMarketNotRunning
+		}
 	}
 
+	if !m.Running() {
+		return sendErr(ErrMarketNotRunning)
+	}
+
+	// If orderRouter is full and blocking, there may just be an actual rush of
+	// orders that will clear out, or we've raced against Run's shutdown defer
+	// and lost.
 	sig := newOrderUpdateSignal(rec)
-	m.orderRouter <- sig
+	select {
+	case m.orderRouter <- sig:
+	default:
+		// If a synchronous send failed, spawn a goroutine with a send timeout.
+		log.Warnf("Incoming order queue full. Attempting time-limited send.")
+		go trySend(sig)
+	}
```

Comment:
I don't love this, but there is a chance that Market.Run has returned in a narrow time window despite best efforts to verify or foresee the market status prior to submission.  Most of the alternatives that I see risk introducing high lock contention to to deal with this relatively rate case, although one idea I might investigate is including in this `select` a channel that is closed when the market stops, which would eliminate the ugly send-with-timeout, but I was hesitant to reverse the semantics of the current `Market.running` channel or to introduce a second channel.

Reply:
These design issues are resolved in https://github.com/decred/dcrdex/pull/287/commits/edbf282d9c9b84cf1b54182b04d8d8133fc42ca4.  Sorry for pushing commits while you're reviewing.

I also decided to simplify the epochPump and resolve a possible deadlock in https://github.com/decred/dcrdex/pull/287/commits/04bbc2e608995f7e7be6a63ffec6a59da5106933